nototools/generate_samples.py

```diff
@@ -0,0 +1,534 @@
+  def __init__(self, patterns, pattern_order):
+    self.patterns = patterns
+    self.pattern_order = pattern_order
+
+  def generate(self, out_file, select_patterns, group, sep, label, sort):
+    if not select_patterns:
+      select_patterns = self.pattern_order
+    else:
+      ok_patterns = []
+      for pattern in select_patterns:
+        if pattern not in self.patterns:
+          print 'No pattern named \'%s\' in %s' % (
+              pattern, ', '.join(self.pattern_order))
+          continue
+        ok_patterns.append(pattern)
+      select_patterns = ok_patterns
+
+    output_lines = []
+    for pattern in select_patterns:
+      self._generate_output(output_lines, pattern, group, sep, label, sort)
+    if not label:
+      # force trailing newline, if we label we already have one
+      output_lines.append('')
+    output_text = '\n'.join(output_lines)
+
+    if out_file:
+      with codecs.open(out_file, 'w', 'utf-8') as f:
+        f.write(output_text)
+    else:
+      print output_text
+
+
+  def _generate_output(self, output_lines, pattern, group, sep, label, sort):
+    pat_list = self.patterns[pattern]
+    if label:
+      output_lines.append(pattern)
+    pat_output = []
+    self._gen_results(pat_list, sort, '', pat_output)
+    if group:
+      output_lines.append(sep.join(pat_output))
+    else:
+      output_lines.extend(pat_output)
+    if label:
+      output_lines.append('')
+
+  def _gen_results(self, pat_list, sort, prefix, samples):
+    if not pat_list:
+      samples.append(prefix)
+      return
+    items = self._get_items(pat_list[0], sort)
+    for item in items:
+      self._gen_results(pat_list[1:], sort, prefix + item, samples)
+
+
+  def _get_items(self, group, sort):
+    if type(group) == tuple:
+      items = []
+      for subgroup in group:
+        for item in self._get_items(subgroup, False):
+          # ensure no duplicates result from union of groups
+          if item not in items:
+            items.append(item)
+    else:
+      items = group
+    return sorted(items) if sort else items
+
+
+# parser utilities
+
+def _strip_comments(definition_lines):
+  """Not as straightforward as usual, because comments can be escaped
+  by backslash, and backslash can escape space."""
+  out_lines = []
+  for line in definition_lines:
+    pos = 0
+    while True:
+      x = line.find('#', pos)
+      if x <= 0:
+        if x == 0:
+          line = ''
+        break
+
+      is_escaped = False
+      if line[x - 1] == '\\':
+        # see how many there are
+        y = x - 2
+        while y >= pos and line[y] == '\\':
+          y -= 1
+        is_escaped = (x - y) % 2 == 0
+
+      if is_escaped:
+        pos = x + 1
+      else:
+        line = line[:x]
+        break
+    out_lines.append(line)
+  return out_lines
+
+
+_ESCAPES = {
+    ' ': '\\x20',
+    '#': '\\x23',
+    ',': '\\x2c',
+    '-': '\\x2d',
+    '\\': '\\x5c',
+    }
+
+def _canonicalize_escapes(definition_lines):
```

Comment:
Yes, I think I can get that to work. I have to make sure to handle '\\ ' properly, so I have to replace '\\' before anything else, so I'll use a list of tuples instead of a map.